FreeRTOS/Test/CMock/event_groups/event_groups_utest.c

```diff
@@ -0,0 +1,608 @@
+#define BIT_0            ( 1 << 0 )
+#define BIT_2            ( 1 << 2 )
+#define BIT_4            ( 1 << 4 )
+#define ALL_SYNC_BITS    ( BIT_0 | BIT_2 | BIT_4 )
+
+/* ===========================  GLOBAL VARIABLES  =========================== */
+
+/**
+ * @brief Global message buffer handle used for tests.
+ */
+static EventGroupHandle_t xEventGroupHandle;
+static List_t xListTemp = { 0 };
+static List_t * pxListTemp = &xListTemp;
+static ListItem_t xListItemDummy = { 0 };
+static ListItem_t * pxListItem_HasTaskBlockOnBit0 = &xListItemDummy;
+
+/* ===========================  EXTERN VARIABLES  =========================== */
+
+/* ==========================  CALLBACK FUNCTIONS =========================== */
+
+void * pvPortMalloc( size_t xSize )
+{
+    return unity_malloc( xSize );
+}
+void vPortFree( void * pv )
+{
+    return unity_free( pv );
+}
+
+/* ============================  Unity Fixtures  ============================ */
+/*! called before each testcase */
+void setUp( void )
+{
+    xEventGroupHandle = NULL;
+
+    pxListTemp->pxIndex = ( ListItem_t * ) &( pxListTemp->xListEnd );
+    pxListTemp->xListEnd.xItemValue = portMAX_DELAY;
+    pxListTemp->xListEnd.pxNext = ( ListItem_t * ) &( pxListTemp->xListEnd );
+    pxListTemp->xListEnd.pxPrevious = ( ListItem_t * ) &( pxListTemp->xListEnd );
+    pxListTemp->uxNumberOfItems = ( UBaseType_t ) 0U;
+
+    pxListItem_HasTaskBlockOnBit0->xItemValue = BIT_0;
+
+    vFakeAssert_Ignore();
+    vFakePortEnterCriticalSection_Ignore();
+    vFakePortExitCriticalSection_Ignore();
+    ulFakePortSetInterruptMaskFromISR_IgnoreAndReturn( 0U );
+    vFakePortClearInterruptMaskFromISR_Ignore();
+
+    /* Track calls to malloc / free */
+    UnityMalloc_StartTest();
+}
+
+/*! called after each testcase */
+void tearDown( void )
+{
+    UnityMalloc_EndTest();
+}
+
+/*! called at the beginning of the whole suite */
+void suiteSetUp()
+{
+}
+
+/*! called at the end of the whole suite */
+int suiteTearDown( int numFailures )
+{
+    return numFailures;
+}
+
+/* ===========================  Static Functions  =========================== */
+
+
+
+/* ==============================  Test Cases  ============================== */
+
+/*!
+ * @brief validate dynamically creating and deleting a new RTOS event group,
+ * @coverage xEventGroupCreate
+ */
+void test_xEventGroupDynamicCreateAndDelete_Success( void )
+{
+    /* Expectation of Function: xEventGroupCreate */
+    vListInitialise_Expect( 0 );
+    vListInitialise_IgnoreArg_pxList();
+    vListInitialise_ReturnThruPtr_pxList( pxListTemp );
+    /* Expectation of Function: vEventGroupDelete */
+    vTaskSuspendAll_Ignore();
+    listCURRENT_LIST_LENGTH_ExpectAnyArgsAndReturn( 1 );
+    vTaskRemoveFromUnorderedEventList_Ignore();
+    listCURRENT_LIST_LENGTH_ExpectAnyArgsAndReturn( 0 );
+    xTaskResumeAll_IgnoreAndReturn( 1 );
+
+    /* API to Test */
+    xEventGroupHandle = xEventGroupCreate();
+    TEST_ASSERT_NOT_EQUAL( NULL, xEventGroupHandle );
+
```

Comment:
You should verify the size of the memory allocation against the size of StaticEventGroup_t.

Reply:
Why?

Reply:
It's good practice to validate inputs to external functions. If you had mocked malloc, you probably would have verified the input parameter right?

Reply:
How to achieve that? Probably I didn't get your point here. There is no way to check if the malloc size is correct in the unit test since it only returns a handler. And the functionally of malloc should be in malloc unit test not in the event_groups unit test?

Reply:
sizeof( EventGroup_t ) is guaranteed to == sizeof( StaticEventGroup_t ).

You could accomplish this with a global variable that you set in your pvPortMalloc() callback.